Fix(AnimatedScrollView): exclude outgoing items in removeAllItems

kazbeksultanov opened the pull request:
<!--
Thanks for filing a pull request!
Reviewers are typically assigned within a week of filing a request.
To learn more about code review, see our documentation on Tree Hygiene: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
-->

Fix: #176362

Issue provided reproducible code related to exception being thrown when removing all items in an `AnimatedList` when some of the item were already being removed but had a long animation duration.

In this PR: The `removeAllItems` method now correctly calculates the range of items to remove, excluding those already undergoing a removal animation. This prevents an assert from triggering when `removeAllItems` is called while other items are still being removed.

<details open>
<summary> Video example after fix </summary>

https://github.com/user-attachments/assets/c7351702-b829-4e35-9978-b95f9f3ae8cd


</details>


<details open>
<summary> Updated reproducible code </summary>

```
import 'package:flutter/material.dart';

void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(debugShowCheckedModeBanner: false, home: AnimatedListExample());
  }
}

class AnimatedListExample extends StatefulWidget {
  const AnimatedListExample({super.key});

  @override
  State<AnimatedListExample> createState() => _AnimatedListExampleState();
}

class _AnimatedListExampleState extends State<AnimatedListExample> {
  final GlobalKey<AnimatedListState> _listKey = GlobalKey<AnimatedListState>();
  final List<int> _items = List<int>.generate(10, (int index) => index);

  void _removeItem(int index) {
    final int removedItem = _items[index];
    _listKey.currentState?.removeItem(
      index,
      (BuildContext context, Animation<double> animation) => SizeTransition(
        sizeFactor: animation,
        child: _buildItem(removedItem, animation, isRemoved: true),
      ),
      // 👇 Long delay so you can press "Remove All" while it’s still animating
      duration: const Duration(seconds: 10),
    );
    _items.removeAt(index);
  }

  void _addItem() {
    final int addingItem = _items.length;
    _listKey.currentState?.insertItem(addingItem, duration: const Duration(seconds: 3));
    _items.add(addingItem);
  }

  void _removeAll() {
    _listKey.currentState?.removeAllItems((BuildContext context, Animation<double> animation) {
      return SizeTransition(
        sizeFactor: animation,
        child: Container(
          color: Colors.red[100],
          child: const ListTile(title: Text('Removing...')),
        ),
      );
    }, duration: const Duration(seconds: 2));
    _items.clear();
  }

  Widget _buildItem(int item, Animation<double> animation, {bool isRemoved = false}) {
    return SizeTransition(
      sizeFactor: animation,
      child: Card(
        margin: const EdgeInsets.symmetric(vertical: 4, horizontal: 8),
        child: ListTile(
          title: Text('Item $item'),
          trailing: !isRemoved
              ? IconButton(
                  icon: const Icon(Icons.delete, color: Colors.red),
                  onPressed: () {
                    final int index = _items.indexOf(item);
                    if (index != -1) {
                      _removeItem(index);
                    }
                  },
                )
              : null,
        ),
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('AnimatedList Example'),
        actions: <Widget>[
          TextButton(onPressed: _addItem, child: const Text('Add')),
          TextButton(
            onPressed: _items.isNotEmpty ? _removeAll : null,
            child: const Text('Clear list'),
          ),
        ],
      ),
      body: AnimatedList(
        key: _listKey,
        initialItemCount: _items.length,
        itemBuilder: (BuildContext context, int index, Animation<double> animation) {
          return _buildItem(_items[index], animation);
        },
      ),
    );
  }
}
```

</details>


## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

**Note**: The Flutter team is currently trialing the use of [Gemini Code Assist for GitHub](https://developers.google.com/gemini-code-assist/docs/review-github-code). Comments from the `gemini-code-assist` bot should not be taken as authoritative feedback from the Flutter team. If you find its comments useful you can update your code accordingly, but if you are unsure or disagree with the feedback, please feel free to wait for a Flutter team member's review for guidance on which automated comments should be addressed.

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

[![talabat.com](https://img.shields.io/badge/talabat.com-contributions-FF5A00?style=flat&logo=flutter&logoColor=white)](https://www.talabat.com) [![Talabat Flutter PRs](https://img.shields.io/badge/Talabat_Flutter_PRs-10%20merged-97ca00?style=flat&logo=flutter&logoColor=white)](https://github.com/search?q=org%3Aflutter+talabat&type=pullrequests)

kazbeksultanov committed:
Fix(AnimatedScrollView): exclude outgoing items in removeAllItems

The `removeAllItems` method now correctly calculates the range of items to remove, excluding those already undergoing a removal animation. This prevents an assert from triggering when `removeAllItems` is called while other items are still being removed.

kazbeksultanov commented:
@chunhtai @chinmaygarde can I ask for review?

justinmc reviewed (commented):
Can you add a test for this?

kazbeksultanov committed:
Merge branch 'master' into animation_scroll_view_remove_all_fix

kazbeksultanov committed:
Fix(flutter): safely handle removeAllItems during item removal

Adds a test case to ensure that calling `removeAllItems` on an `AnimatedList` while another item is already in the process of being removed executes safely and correctly clears the list.

kazbeksultanov commented:
@justinmc added test for the exact scenario that I am testing. Waiting for review

justinmc commented on packages/flutter/lib/src/widgets/animated_scroll_view.dart:
```diff
@@ -1437,9 +1439,13 @@ abstract class _SliverAnimatedMultiBoxAdaptorState<T extends _SliverAnimatedMult
+    assert(_itemsCount >= 0);
+    assert(_itemsCount - _outgoingItems.length >= 0);
+    final int notOutGoingTotalCount = _itemsCount - _outgoingItems.length;
+    for (int i = notOutGoingTotalCount - 1; i >= 0; i--) {
```
I like this as well, it avoids the negative in the name.

justinmc commented on packages/flutter/test/widgets/animated_list_test.dart:
```diff
@@ -107,6 +107,73 @@ void main() {
+      ),
+    );
+
+    // Check that one AnimatedList with 2 items (item 0, item 1)
```
Nit: Periods at the ends of these comments.

justinmc reviewed (approved):
LGTM with nits 👍

kazbeksultanov committed:
refactor: improve AnimatedList.removeAllItems documentation and implementation

The documentation for `AnimatedList.removeAllItems` is clarified to better explain its behavior, specifying that it removes all items, including those being inserted, but excludes those already being removed.

The implementation is updated to use a more descriptive variable name, `visibleItemCount`, for better code readability.

Additionally, periods have been added to comments in the corresponding test file for grammatical correctness.

kazbeksultanov committed:
Merge branch 'master' into animation_scroll_view_remove_all_fix

kazbeksultanov committed:
Merge branch 'master' into animation_scroll_view_remove_all_fix

kazbeksultanov committed:
Merge branch 'master' into animation_scroll_view_remove_all_fix

kazbeksultanov committed:
Merge branch 'master' into animation_scroll_view_remove_all_fix

chunhtai reviewed (approved):
LGTM

chunhtai committed:
Merge branch 'master' into animation_scroll_view_remove_all_fix

chunhtai commented:
Can't tell whats wrong from google test log, rerun to see if it passes

kazbeksultanov committed:
Merge branch 'master' into animation_scroll_view_remove_all_fix

justinmc reviewed (approved):
LGTM, will autosubmit
